File: Task/Numbers-with-equal-rises-and-falls/Haskell/numbers-with-equal-rises-and-falls.hs
import Data.Char

pairs :: [a] -> [(a,a)]
pairs (a:b:as) = (a,b):pairs (b:as)
pairs _        = []

riseEqFall :: Int -> Bool
riseEqFall n = rel (>) digitPairs == rel (<) digitPairs
    where rel r = sum . map (fromEnum . uncurry r)
          digitPairs = pairs $ map digitToInt $ show n

a296712 :: [Int]
a296712 = [n | n <- [1..], riseEqFall n]

main :: IO ()
main = do
	putStrLn "The first 200 numbers are: "
	putStrLn $ unwords $ map show $ take 200 a296712
	putStrLn ""
	putStr "The 10,000,000th number is: "
	putStrLn $ show $ a296712 !! 9999999

File: Task/Numbers-with-equal-rises-and-falls/PascalABC.NET/numbers-with-equal-rises-and-falls.pas
function a296712(): sequence of integer;
begin
  foreach var n in 1.step do
    if n.ToString
        .Pairwise((x, y) -> (if x > y then 1 else if x < y then -1 else 0))
        .Sum = 0 then
      yield n;
end;

begin
  println('The first 200 numbers are:');
  a296712.Take(200).Print;
  println;
  println;
  println('The 10,000,000th number is:');
  a296712.ElementAt(10_000_000 - 1).Println;
end.
